File: swift/TestRSA.swift
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

//
//  TestRSA.swift
//  
//
//  Created by Michael Scott on 25/06/2015.
//  Copyright (c) 2015 Michael Scott. All rights reserved.
//

import Foundation
import clint   // comment out for Xcode

public func TestRSA()
{
    let RFS=RSA.RFS;

    var message="Hello World\n"

    let pub=rsa_public_key(ROM.FFLEN);
    let priv=rsa_private_key(ROM.HFLEN);

    var ML=[UInt8](count:RFS,repeatedValue:0)
    var C=[UInt8](count:RFS,repeatedValue:0)
    var RAW=[UInt8](count:100,repeatedValue:0)

    let rng=RAND()

    rng.clean();
    for var i=0;i<100;i++ {RAW[i]=UInt8(i)}

    rng.seed(100,RAW);

    print("Generating public/private key pair");
    RSA.KEY_PAIR(rng,65537,priv,pub);

    let M=[UInt8](message.utf8)
    print("Encrypting test string\n");
    let E=RSA.OAEP_ENCODE(M,rng,nil); /* OAEP encode message m to e  */

    RSA.ENCRYPT(pub,E,&C);     /* encrypt encoded message */
    print("Ciphertext= 0x", terminator: ""); RSA.printBinary(C);

    print("Decrypting test string\n");
    RSA.DECRYPT(priv,C,&ML);
    var MS=RSA.OAEP_DECODE(nil,&ML); /* OAEP encode message m to e  */

    message=""
    for var i=0;i<MS.count;i++
    {
        message+=String(UnicodeScalar(MS[i]))
    }
    print(message);

    RSA.PRIVATE_KEY_KILL(priv);
}

TestRSA() // comment out for Xcode

File: cs/TestRSA.cs
using System;

/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

/* test driver and function exerciser for RSA API Functions */

public class TestRSA
{

	public static void Main(string[] args)
	{
		int i;
		int RFS = RSA.RFS;

		string message = "Hello World\n";

		rsa_public_key pub = new rsa_public_key(ROM.FFLEN);
		rsa_private_key priv = new rsa_private_key(ROM.HFLEN);

		sbyte[] ML = new sbyte[RFS];
		sbyte[] C = new sbyte[RFS];
		sbyte[] RAW = new sbyte[100];

		RAND rng = new RAND();

		rng.clean();
		for (i = 0;i < 100;i++)
		{
			RAW[i] = (sbyte)(i);
		}

		rng.seed(100,RAW);
//for (i=0;i<10;i++)
//{
		Console.WriteLine("Generating public/private key pair");
		RSA.KEY_PAIR(rng,65537,priv,pub);

		sbyte[] M = message.GetBytes();
		Console.Write("Encrypting test string\n");
		sbyte[] E = RSA.OAEP_ENCODE(M,rng,null); // OAEP encode message M to E

		RSA.ENCRYPT(pub,E,C); // encrypt encoded message
		Console.Write("Ciphertext= 0x");
		RSA.printBinary(C);

		Console.Write("Decrypting test string\n");
		RSA.DECRYPT(priv,C,ML);
		sbyte[] MS = RSA.OAEP_DECODE(null,ML); // OAEP decode message

		message = StringHelperClass.NewString(MS);
		Console.Write(message);
//}
		RSA.PRIVATE_KEY_KILL(priv);
	}
}
